File: CharacterManagement.Domain/Entity/Deity.cs
namespace Pathfinder.CharacterManagement.Domain.Entity;

public enum DivineFont
{
    Heal,
    Harm
}

public enum DivineSanctification
{
    Holy,
    Unholy
}

public enum FavoredWeaponCategory
{
    Simple,
    Martial,
    Advanced,
    Unarmed
}

public sealed record DeityFavoredWeapon(
    string Id,
    string Name,
    FavoredWeaponCategory Category );

public sealed record DeityGrantedSpell(
    int Rank,
    string Id,
    string Name );

public sealed class Deity
{
    public string Id { get; }
    public string Name { get; }
    public SourceReference Source { get; }
    public bool CanGrantClericPowers { get; }
    public string? DivineSkillId { get; }
    public IReadOnlyList<DeityFavoredWeapon> FavoredWeapons { get; }
    public IReadOnlyList<DivineFont> DivineFontOptions { get; }
    public IReadOnlyList<DivineSanctification> SanctificationOptions { get; }
    public DivineSanctification? RequiredSanctification { get; }
    public IReadOnlyList<string> PrimaryDomainIds { get; }
    public IReadOnlyList<DeityGrantedSpell> GrantedSpells { get; }
    public IReadOnlyList<ProficiencyGrant> ProficiencyGrants { get; }
    public string DivineSkillGrantId => $"{Id}.skill.divine";

    public Deity(
        string id,
        string name,
        SourceReference source,
        bool canGrantClericPowers,
        string? divineSkillId,
        IReadOnlyList<DeityFavoredWeapon> favoredWeapons,
        IReadOnlyList<DivineFont> divineFontOptions,
        IReadOnlyList<DivineSanctification> sanctificationOptions,
        DivineSanctification? requiredSanctification,
        IReadOnlyList<string> primaryDomainIds,
        IReadOnlyList<DeityGrantedSpell> grantedSpells )
    {
        if ( String.IsNullOrWhiteSpace( id ) ||
             !id.StartsWith( "deity.", StringComparison.Ordinal ) )
        {
            throw new ArgumentException( "Deity id must use the 'deity.' prefix.", nameof( id ) );
        }

        if ( String.IsNullOrWhiteSpace( name ) )
        {
            throw new ArgumentException( "Deity name cannot be empty.", nameof( name ) );
        }

        ArgumentNullException.ThrowIfNull( source );
        ArgumentNullException.ThrowIfNull( favoredWeapons );
        ArgumentNullException.ThrowIfNull( divineFontOptions );
        ArgumentNullException.ThrowIfNull( sanctificationOptions );
        ArgumentNullException.ThrowIfNull( primaryDomainIds );
        ArgumentNullException.ThrowIfNull( grantedSpells );

        if ( canGrantClericPowers &&
             ( String.IsNullOrWhiteSpace( divineSkillId ) ||
               favoredWeapons.Count == 0 ||
               divineFontOptions.Count == 0 ||
               primaryDomainIds.Count == 0 ) )
        {
            throw new ArgumentException( "A Cleric deity must define all devotee benefits." );
        }

        if ( !canGrantClericPowers &&
             ( !String.IsNullOrWhiteSpace( divineSkillId ) ||
               favoredWeapons.Count > 0 ||
               divineFontOptions.Count > 0 ||
               sanctificationOptions.Count > 0 ||
               requiredSanctification.HasValue ||
               primaryDomainIds.Count > 0 ||
               grantedSpells.Count > 0 ) )
        {
            throw new ArgumentException( "A non-Cleric faith cannot define devotee benefits." );
        }

        if ( requiredSanctification.HasValue &&
             !sanctificationOptions.Contains( requiredSanctification.Value ) )
        {
            throw new ArgumentException( "Required sanctification must be an allowed option." );
        }

        Id = id.Trim();
        Name = name.Trim();
        Source = source;
        CanGrantClericPowers = canGrantClericPowers;
        DivineSkillId = divineSkillId?.Trim();
        FavoredWeapons = favoredWeapons.ToArray();
        DivineFontOptions = divineFontOptions.Distinct().ToArray();
        SanctificationOptions = sanctificationOptions.Distinct().ToArray();
        RequiredSanctification = requiredSanctification;
        PrimaryDomainIds = primaryDomainIds.Distinct( StringComparer.Ordinal ).ToArray();
        GrantedSpells = grantedSpells.ToArray();
        ProficiencyGrants = FavoredWeapons
            .Select( weapon => new ProficiencyGrant(
                new ProficiencyTarget(
                    $"proficiency.attack.{weapon.Id}",
                    weapon.Name,
                    ProficiencyCategory.Attack ),
                ProficiencyRank.Trained,
                $"{Id}.proficiency.{weapon.Id}" ) )
            .ToArray();
    }
}

File: pathfinder.frontend/src/features/character-creation/deity.ts
import type { Proficiency } from '@/features/characters/api'
import type {
  CharacterClass,
  Deity,
  DivineFont,
  DivineSanctification,
  Skill,
} from './api'

export function requiresDeitySkillReplacement(
  deity: Deity | null,
  backgroundSkillIds: string[],
): boolean {
  const divineSkillId = deity?.divineSkillId ?? null
  return divineSkillId !== null && backgroundSkillIds.includes(divineSkillId)
}

export function isDeityChoiceComplete(
  characterClass: CharacterClass | null,
  deity: Deity | null,
  divineFont: DivineFont | null,
  sanctification: DivineSanctification | null,
  replacementSkillId: string | null,
  backgroundSkillIds: string[],
  skills: Skill[],
): boolean {
  if (characterClass?.id !== 'class.cleric') {
    return deity === null && divineFont === null && sanctification === null && replacementSkillId === null
  }
  if (!deity?.canGrantClericPowers || !divineFont || !deity.divineFontOptions.includes(divineFont)) return false
  if (deity.requiredSanctification && sanctification !== deity.requiredSanctification) return false
  if (sanctification && !deity.sanctificationOptions.includes(sanctification)) return false

  const requiresReplacement = requiresDeitySkillReplacement(deity, backgroundSkillIds)
  if (!requiresReplacement) return replacementSkillId === null
  return Boolean(
    replacementSkillId &&
      !backgroundSkillIds.includes(replacementSkillId) &&
      skills.some((skill) => skill.id === replacementSkillId),
  )
}

export function getDeityProficiencies(deity: Deity | null): Proficiency[] {
  return deity?.favoredWeapons.map((weapon) => ({
    targetId: `proficiency.attack.${weapon.id}`,
    name: weapon.name,
    category: 'Attack',
    rank: 'Trained',
    sourceGrantId: `${deity.id}.proficiency.${weapon.id}`,
  })) ?? []
}
